File: rust-src/commands/stats.rs
//! Stats command - show detailed statistics.

use anyhow::Result;
use chrono::{Datelike, NaiveDate};
use std::collections::HashSet;

use crate::config::{get_claude_jsonl_files, get_db_path};
use crate::data::parse_jsonl_file;
use crate::storage::{get_database_stats, save_snapshot};


/// Run the stats command.
pub fn run(fast: bool) -> Result<()> {
    let db_path = get_db_path();

    // Check for fast mode with no database
    if fast && !db_path.exists() {
        eprintln!("Error: Cannot use --fast flag without existing database.");
        eprintln!("Run 'ccg stats' (without --fast) first to create the database.");
        return Ok(());
    }

    // Update data unless in fast mode
    if !fast {
        println!("Updating database...");
        if let Ok(jsonl_files) = get_claude_jsonl_files() {
            if !jsonl_files.is_empty() {
                let mut all_records = Vec::new();
                for file in &jsonl_files {
                    if let Ok(records) = parse_jsonl_file(file) {
                        all_records.extend(records);
                    }
                }
                if !all_records.is_empty() {
                    let _ = save_snapshot(&all_records, &db_path);
                }
            }
        }
    } else {
        println!("Fast mode: Reading from database...\n");
    }

    // Get stats from database
    let stats = get_database_stats(&db_path)?;

    if stats.total_records == 0 && stats.total_prompts == 0 {
        println!("No historical data found. Run 'ccg update usage' to start tracking.");
        return Ok(());
    }

    // Header
    println!("\n{}", "=".repeat(60));
    println!("{:^60}", "Claude Code Usage Statistics");
    println!("{}\n", "=".repeat(60));

    // Summary Statistics
    println!("SUMMARY");
    println!("{}", "-".repeat(40));
    println!("  Total Tokens:        {:>15}", format_number(stats.total_tokens));
    println!("  Total Prompts:       {:>15}", format_number(stats.total_prompts));
    println!("  Total Responses:     {:>15}", format_number(stats.total_responses));
    println!("  Total Sessions:      {:>15}", format_number(stats.total_sessions));
    println!("  Days Tracked:        {:>15}", format_number(stats.total_days));

    if let (Some(oldest), Some(newest)) = (&stats.oldest_date, &stats.newest_date) {
        println!("  Date Range:          {} to {}", oldest, newest);
    }

    // Cost Analysis
    if stats.total_cost > 0.0 {
        println!("\nCOST ANALYSIS");
        println!("{}", "-".repeat(40));
        println!("  Est. Cost (API):     ${:>14}", format_currency(stats.total_cost));

        // Calculate months covered
        if let (Some(oldest), Some(newest)) = (&stats.oldest_date, &stats.newest_date) {
            if let (Ok(start), Ok(end)) = (
                NaiveDate::parse_from_str(oldest, "%Y-%m-%d"),
                NaiveDate::parse_from_str(newest, "%Y-%m-%d"),
            ) {
                let mut months = HashSet::new();
                let mut current = start;
                while current <= end {
                    months.insert((current.year(), current.month()));
                    current = current.checked_add_days(chrono::Days::new(1)).unwrap_or(end);
                }
                let num_months = months.len().max(1);
                let plan_cost = num_months as f64 * 200.0;
                let savings = stats.total_cost - plan_cost;

                println!(
                    "  Plan Cost:           ${:>14} ({} month{} @ $200/mo)",
                    format_currency(plan_cost),
                    num_months,
                    if num_months > 1 { "s" } else { "" }
                );

                if savings > 0.0 {
                    println!("  You Saved:           ${:>14} (vs API)", format_currency(savings));
                } else {
                    println!("  Plan Costs More:     ${:>14}", format_currency(savings.abs()));
                    println!("  [Light usage - API would be cheaper]");
                }
            }
        }
    }

    // Averages
    println!("\nAVERAGES");
    println!("{}", "-".repeat(40));
    let avg_per_session = if stats.total_sessions > 0 {
        stats.total_tokens / stats.total_sessions
    } else {
        0
    };
    let avg_per_response = if stats.total_responses > 0 {
        stats.total_tokens / stats.total_responses
    } else {
        0
    };
    println!("  Tokens per Session:  {:>15}", format_number(avg_per_session));
    println!("  Tokens per Response: {:>15}", format_number(avg_per_response));

    if stats.total_cost > 0.0 && stats.total_sessions > 0 {
        let cost_per_session = stats.total_cost / stats.total_sessions as f64;
        let cost_per_response = if stats.total_responses > 0 {
            stats.total_cost / stats.total_responses as f64
        } else {
            0.0
        };
        println!("  Cost per Session:    ${:>14}", format_currency(cost_per_session));
        println!("  Cost per Response:   ${:>14}", format_currency_4(cost_per_response));
    }

    // Tokens by Model
    if !stats.tokens_by_model.is_empty() {
        println!("\nUSAGE BY MODEL");
        println!("{}", "-".repeat(60));

        let mut models: Vec<_> = stats.tokens_by_model.iter().collect();
        models.sort_by(|a, b| b.1.cmp(a.1));

        for (model, tokens) in models {
            let percentage = if stats.total_tokens > 0 {
                (*tokens as f64 / stats.total_tokens as f64) * 100.0
            } else {
                0.0
            };
            let cost = stats.cost_by_model.get(model).copied().unwrap_or(0.0);

            if cost > 0.0 {
                println!(
                    "  {:30} {:>12} ({:5.1}%) ${:>10}",
                    model,
                    format_number(*tokens),
                    percentage,
                    format_currency(cost)
                );
            } else {
                println!(
                    "  {:30} {:>12} ({:5.1}%)",
                    model,
                    format_number(*tokens),
                    percentage
                );
            }
        }
    }

    // Database Info
    println!("\n{}", "-".repeat(60));
    println!("Database: ~/.claude/usage/usage_history.db");
    if stats.total_records > 0 {
        println!("Detail records: {} (full analytics mode)", format_number(stats.total_records));
    } else {
        println!("Storage mode: aggregate (daily totals only)");
    }

    Ok(())
}


/// Format a number with commas.
fn format_number(n: i64) -> String {
    let s = n.to_string();
    let mut result = String::new();
    let chars: Vec<char> = s.chars().collect();

    for (i, c) in chars.iter().enumerate() {
        if i > 0 && (chars.len() - i) % 3 == 0 {
            result.push(',');
        }
        result.push(*c);
    }

    result
}


/// Format currency with 2 decimal places and commas.
fn format_currency(n: f64) -> String {
    let formatted = format!("{:.2}", n);
    let parts: Vec<&str> = formatted.split('.').collect();
    let integer_part = parts[0];
    let decimal_part = parts.get(1).unwrap_or(&"00");

    let mut result = String::new();
    let chars: Vec<char> = integer_part.chars().collect();

    for (i, c) in chars.iter().enumerate() {
        if i > 0 && (chars.len() - i) % 3 == 0 {
            result.push(',');
        }
        result.push(*c);
    }

    format!("{}.{}", result, decimal_part)
}


/// Format currency with 4 decimal places.
fn format_currency_4(n: f64) -> String {
    format!("{:.4}", n)
}

File: src/commands/stats.py
#region Imports
import sys
from datetime import datetime

from rich.console import Console

from src.commands.limits import capture_limits
from src.commands.update_usage import ingest_token_usage
from src.config.user_config import get_tracking_mode
from src.storage import api
from src.storage.api import (
    get_database_stats,
    get_text_analysis_stats,
    save_limits_snapshot,
)
#endregion


#region Functions


def run(console: Console, fast: bool = False, force: bool = False) -> None:
    """
    Show statistics about the historical database.

    Displays comprehensive statistics including:
    - Summary: total tokens, prompts, responses, sessions, days tracked
    - Cost analysis: estimated API costs vs Max Plan costs
    - Averages: tokens per session/response, cost per session/response
    - Text analysis: prompt length, politeness markers, phrase counts
    - Usage by model: token distribution across different models

    Args:
        console: Rich console for output
        fast: Skip updates, read directly from database (default: False)
        force: Force re-parse all files, ignoring incremental cache (default: False)
    """
    # Check for flags in sys.argv for backward compatibility
    fast_mode = fast or "--fast" in sys.argv
    force_mode = force or "--force" in sys.argv

    # Check if database exists when using --fast
    if fast_mode and not api.current_db_path().exists():
        console.print("[red]Error: Cannot use --fast flag without existing database.[/red]")
        console.print("[yellow]Run 'ccg stats' (without --fast) first to create the database.[/yellow]")
        return

    # If fast mode, show warning with last update timestamp
    if fast_mode:
        db_stats_temp = get_database_stats()
        if db_stats_temp.get("newest_timestamp"):
            # Format ISO timestamp to be more readable
            timestamp_str = db_stats_temp["newest_timestamp"]
            try:
                dt = datetime.fromisoformat(timestamp_str)
                formatted_time = dt.strftime("%Y-%m-%d %H:%M:%S")
                console.print(f"[bold red]⚠ Fast mode: Reading from last update ({formatted_time})[/bold red]\n")
            except (ValueError, AttributeError):
                console.print(f"[bold red]⚠ Fast mode: Reading from last update ({timestamp_str})[/bold red]\n")
        else:
            console.print("[bold red]⚠ Fast mode: Reading from database (no timestamp available)[/bold red]\n")

    # Update data unless in fast mode
    if not fast_mode:
        # Step 1: Ingestion - parse JSONL and save to DB (incremental)
        if force_mode:
            console.print("[yellow]Force mode: reparsing all files (this may take a moment)[/yellow]")
        with console.status("[bold #ff8800]Updating changed files...", spinner="dots", spinner_style="#ff8800"):
            ingest_token_usage(console, force=force_mode, verbose=False)

        # Step 2: Update limits data (if enabled and working)
        # NOTE: Limits tracking is currently disabled due to Claude Code format changes
        tracking_mode = get_tracking_mode()
        if tracking_mode in ["both", "limits"]:
            limits = capture_limits()
            if limits and "error" not in limits:
                with console.status("[bold #ff8800]Updating usage limits...", spinner="dots", spinner_style="#ff8800"):
                    save_limits_snapshot(
                        session_pct=limits["session_pct"],
                        week_pct=limits["week_pct"],
                        opus_pct=limits["opus_pct"],
                        session_reset=limits["session_reset"],
                        week_reset=limits["week_reset"],
                        opus_reset=limits["opus_reset"],
                    )
            # Silently skip if disabled

    # Step 3: Display stats from DB
    db_stats = get_database_stats()

    if db_stats["total_records"] == 0 and db_stats["total_prompts"] == 0:
        console.print("[yellow]No historical data found. Run ccg usage to start tracking.[/yellow]")
        return

    console.print("[bold cyan]Claude Code Usage Statistics[/bold cyan]\n")

    # Summary Statistics
    console.print("[bold]Summary[/bold]")
    console.print(f"  Total Tokens:        {db_stats['total_tokens']:>15,}")
    console.print(f"  Total Prompts:       {db_stats['total_prompts']:>15,}")
    console.print(f"  Total Responses:     {db_stats['total_responses']:>15,}")
    console.print(f"  Total Sessions:      {db_stats['total_sessions']:>15,}")
    console.print(f"  Days Tracked:        {db_stats['total_days']:>15,}")
    console.print(f"  Date Range:          {db_stats['oldest_date']} to {db_stats['newest_date']}")

    # Cost Summary (if using API pricing)
    if db_stats['total_cost'] > 0:
        # Calculate actual months covered from date range
        start_date = datetime.strptime(db_stats['oldest_date'], "%Y-%m-%d")
        end_date = datetime.strptime(db_stats['newest_date'], "%Y-%m-%d")

        # Count unique months covered
        months_covered = set()
        current = start_date
        while current <= end_date:
            months_covered.add((current.year, current.month))
            # Move to next month
            if current.month == 12:
                current = current.replace(year=current.year + 1, month=1, day=1)
            else:
                current = current.replace(month=current.month + 1, day=1)

        num_months = len(months_covered)
        plan_cost = num_months * 200.0  # $200/month Max Plan
        savings = db_stats['total_cost'] - plan_cost

        console.print(f"\n[bold]Cost Analysis[/bold]")
        console.print(f"  Est. Cost (if using API): ${db_stats['total_cost']:>10,.2f}")
        console.print(f"  Plan Cost:           ${plan_cost:>14,.2f} ({num_months} month{'s' if num_months > 1 else ''} @ $200/mo)")

        if savings > 0:
            console.print(f"  You Saved:           ${savings:>14,.2f} (vs API)")
        else:
            overpaid = abs(savings)
            console.print(f"  Plan Costs More:     ${overpaid:>14,.2f}")
            console.print(f"  [dim]Light usage - API would be cheaper[/dim]")

    # Averages
    console.print(f"\n[bold]Averages[/bold]")
    console.print(f"  Tokens per Session:  {db_stats['avg_tokens_per_session']:>15,}")
    console.print(f"  Tokens per Response: {db_stats['avg_tokens_per_response']:>15,}")
    if db_stats['total_cost'] > 0:
        console.print(f"  Cost per Session:    ${db_stats['avg_cost_per_session']:>14,.2f}")
        console.print(f"  Cost per Response:   ${db_stats['avg_cost_per_response']:>14,.4f}")

    # Text Analysis (from current JSONL files)
    text_stats = get_text_analysis_stats()

    if text_stats["avg_user_prompt_chars"] > 0:
        console.print(f"\n[bold]Text Analysis[/bold]")
        console.print(f"  Avg Prompt Length:   {text_stats['avg_user_prompt_chars']:>15,} chars")
        console.print(f"  User Swears:         {text_stats['user_swears']:>15,}")
        console.print(f"  Claude Swears:       {text_stats['assistant_swears']:>15,}")
        console.print(f"  User Thanks:         {text_stats['user_thanks']:>15,}")
        console.print(f"  User Please:         {text_stats['user_please']:>15,}")
        console.print(f"  Claude \"Perfect!\"/\"Excellent!\": {text_stats['perfect_count']:>10,}")
        console.print(f"  Claude \"You're absolutely right!\": {text_stats['absolutely_right_count']:>6,}")

    # Tokens by Model
    if db_stats["tokens_by_model"]:
        console.print(f"\n[bold]Usage by Model[/bold]")
        for model, tokens in db_stats["tokens_by_model"].items():
            percentage = (tokens / db_stats['total_tokens'] * 100) if db_stats['total_tokens'] > 0 else 0
            cost = db_stats["cost_by_model"].get(model, 0.0)
            if cost > 0:
                console.print(f"  {model:30s} {tokens:>15,} ({percentage:5.1f}%) ${cost:>10,.2f}")
            else:
                console.print(f"  {model:30s} {tokens:>15,} ({percentage:5.1f}%)")

    # Database Info
    console.print(f"\n[dim]Database: {api.current_db_path()}[/dim]")
    if db_stats["total_records"] > 0:
        console.print(f"[dim]Detail records: {db_stats['total_records']:,} (full analytics mode)[/dim]")
    else:
        console.print(f"[dim]Storage mode: aggregate (daily totals only)[/dim]")


def run_remote(console: Console) -> None:
    """
    Show statistics from the remote DuckDB server (cross-device aggregate).
    """
    try:
        from src.storage.quack_remote import get_database_stats as remote_stats

        with console.status("[bold #ff8800]Connecting to remote...", spinner="dots", spinner_style="#ff8800"):
            db_stats = remote_stats()

        if db_stats["total_records"] == 0 and db_stats["total_prompts"] == 0:
            console.print("[yellow]No data on remote. Push first with: ccg sync push[/yellow]")
            return

        console.print("[bold cyan]Claude Code Usage Statistics (Remote)[/bold cyan]\n")

        console.print("[bold]Summary[/bold]")
        console.print(f"  Total Tokens:        {db_stats['total_tokens']:>15,}")
        console.print(f"  Total Prompts:       {db_stats['total_prompts']:>15,}")
        console.print(f"  Total Responses:     {db_stats['total_responses']:>15,}")
        console.print(f"  Total Sessions:      {db_stats['total_sessions']:>15,}")
        console.print(f"  Days Tracked:        {db_stats['total_days']:>15,}")
        console.print(f"  Date Range:          {db_stats['oldest_date']} to {db_stats['newest_date']}")

        if db_stats['total_cost'] > 0:
            start_date = datetime.strptime(db_stats['oldest_date'], "%Y-%m-%d")
            end_date = datetime.strptime(db_stats['newest_date'], "%Y-%m-%d")
            months_covered = set()
            current = start_date
            while current <= end_date:
                months_covered.add((current.year, current.month))
                if current.month == 12:
                    current = current.replace(year=current.year + 1, month=1, day=1)
                else:
                    current = current.replace(month=current.month + 1, day=1)
            num_months = len(months_covered)
            plan_cost = num_months * 200.0
            savings = db_stats['total_cost'] - plan_cost

            console.print(f"\n[bold]Cost Analysis[/bold]")
            console.print(f"  Est. Cost (if using API): ${db_stats['total_cost']:>10,.2f}")
            console.print(f"  Plan Cost:           ${plan_cost:>14,.2f} ({num_months} month{'s' if num_months > 1 else ''} @ $200/mo)")
            if savings > 0:
                console.print(f"  You Saved:           ${savings:>14,.2f} (vs API)")
            else:
                console.print(f"  Plan Costs More:     ${abs(savings):>14,.2f}")

        console.print(f"\n[bold]Averages[/bold]")
        console.print(f"  Tokens per Session:  {db_stats['avg_tokens_per_session']:>15,}")
        console.print(f"  Tokens per Response: {db_stats['avg_tokens_per_response']:>15,}")
        if db_stats['total_cost'] > 0:
            console.print(f"  Cost per Session:    ${db_stats['avg_cost_per_session']:>14,.2f}")
            console.print(f"  Cost per Response:   ${db_stats['avg_cost_per_response']:>14,.4f}")

        if db_stats["tokens_by_model"]:
            console.print(f"\n[bold]Usage by Model[/bold]")
            for model, tokens in db_stats["tokens_by_model"].items():
                pct = (tokens / db_stats['total_tokens'] * 100) if db_stats['total_tokens'] > 0 else 0
                cost = db_stats["cost_by_model"].get(model, 0.0)
                if cost > 0:
                    console.print(f"  {model:30s} {tokens:>15,} ({pct:5.1f}%) ${cost:>10,.2f}")
                else:
                    console.print(f"  {model:30s} {tokens:>15,} ({pct:5.1f}%)")

        console.print(f"\n[dim]Source: remote (cross-device aggregate)[/dim]")

    except ImportError:
        console.print("[red]DuckDB not installed. Install with: uv pip install claude-goblin[duckdb][/red]")
    except RuntimeError as e:
        console.print(f"[red]Remote connection failed: {e}[/red]")
    except Exception as e:
        console.print(f"[red]Error: {e}[/red]")


#endregion
